### PR TITLE
Fixed file system iterator, particle orientation and added reverse string hash lookup

### PR DESCRIPTION
**Pull request by jankrassnigg**

### CODE REVIEW

File: Data/Base/Shaders/Particles/ParticleCommonVS.h [hunk 1]
```diff
@@ -60,14 +60,17 @@ Quad CalcQuadOutputPositionWithAlignedAxis(uint vertexIndex, float3 inPosition,
...
   float3 offsetRight = orthoDir * ((QuadTexCoords[vertexIndex].x - 0.5) * inSize);
   float3 offsetUp = axisDir * ((1.0 - QuadTexCoords[vertexIndex].y) * inSize * stretch);
 
   Quad quad;
-  quad.worldPosition = mul(ObjectToWorldMatrix, float4(inPosition + offsetRight + offsetUp, 1));
+  quad.worldPosition = mul(ObjectToWorldMatrix, float4(inPosition, 1));
+  quad.worldPosition.xyz += offsetRight + offsetUp;
   quad.screenPosition = mul(GetWorldToScreenMatrix(), quad.worldPosition);
 
   float3 centerNormal = cross(axisDir, orthoDir);
```
Comment (C-Core) [inline Data/Base/Shaders/Particles/ParticleCommonVS.h:76]: The normal is now wrong since axisDir and orthoDir and the offsets are already in world space.

**Review by C-Core — CHANGES_REQUESTED** [review]

File: Data/Base/Shaders/Particles/ParticleCommonVS.h (hunk 1 above)
Comment (jankrassnigg) [reply Data/Base/Shaders/Particles/ParticleCommonVS.h:76]: You mean 'centerNormal' ? I don't see that.

As far as I can tell, the code originally assumed that axisDir and orthoDir are in world-space, but they only where that, when the effect wasn't rotated (therefore it always looked right in the particle editor).

Now they are always in world space and I think the rest of the calculation doesn't need to change.

Comment (C-Core) [reply Data/Base/Shaders/Particles/ParticleCommonVS.h:76]: The normals are transformed into world space 2 lines below that. Github doesn't allow me to comment on that line :(
Now the normals are double transformed.

Comment (jankrassnigg) [reply Data/Base/Shaders/Particles/ParticleCommonVS.h:76]: Good catch. Fixed it.

**Review by C-Core — APPROVED** [review]

**Pull request merged**